src/ontodia/widgets/node.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+
+import { FatClassModel } from '../diagram/elements';
+import { formatLocalizedLabel } from '../diagram/model';
+import { TreeNodes } from './treeNodes';
+
+export interface NodeTreeProps {
+    node: FatClassModel;
+    lang?: Readonly<string> | undefined;
+    searchResult?: Array<FatClassModel> | undefined;
+    searchString?: string | undefined;
+    onClassSelected: (classId: string) => void;
+    onDragDrop?: (e: DragEvent, paperPosition: { x: number; y: number; }) => void;
+}
+
+export interface ClassTreeState {
+    expanded?: Boolean | undefined;
+    bgColor?: string | undefined;
+    mapClassIcons?: { [typeId: string]: string } | undefined;
+}
+
+const CLASS_NAME = 'ontodia-class-tree';
+
+export class Node extends React.PureComponent<NodeTreeProps, ClassTreeState> {
+    constructor(props: NodeTreeProps) {
+        super(props);
+        let resultEmpty = !(this.props.searchResult && this.props.searchResult.length !== 0);
+        if (resultEmpty) {
+            this.state = { expanded: false };
+        } else {
+            this.state = { expanded: true };
+        }
+
+        /* customization 
+        let iconMap: { [typeId: string]: string } = {};
+        iconMap['http://ailab.ifmo.ru/dialog/tv/schema#PropertyAsGoods'] = 'custom-parent-tree-icon';
+        iconMap['http://www.w3.org/2002/07/owl#Thing'] = 'custom-parent-tree-icon';
+        iconMap['http://www.w3.org/1999/02/22-rdf-syntax-ns#Property'] = 'custom-child-tree-icon';
+        if (Object.keys(iconMap).length !== 0) {
+            if (resultEmpty) {
+                this.state = { mapClassIcons: iconMap, expanded: false };
+            } else {
+                this.state = { mapClassIcons: iconMap, expanded: true };
+            }
+        } */
+        this.toggle = this.toggle.bind(this);
+        this.showInstances = this.showInstances.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps: NodeTreeProps) {
+        const { searchResult } = nextProps;
+        if (searchResult !== this.props.searchResult) {
+            this.setState({ expanded: false });
+        }
+        if (searchResult && searchResult.length !== 0) {
+            this.setState({ expanded: true });
+        }
+        if (this.state.bgColor === 'rgb(190,235,255)') {
```

Comment:
This is very unstable condition: any external update from parent component will reset the color.

Reply:
Sure.
Corrected on `if (this.state.bgColor !== undefined)`

Reply:
I mean, the issue is `componentWillReceiveProps()` could be called at any update from parent component, so `this.state.bgColor` would be set to `undefined` just by accident.

Reply:
I understand the intent, but the implementation is fragile:
> If you select one element, all the others will reset the background color

The issue is this code will run not specifically on selection but on **every** update from any parent components.

Reply:
>The issue is this code will run not specifically on selection but on every update from any parent components.

I tried to do this on pure css, but it did not work the way I want it. Examples: `https://ruseller.com/lessons.php?id=1600`

The approach that I have chosen suits me. Why? `Render()` is called whenever `Proprs` or `State` changes.
When changing the `classesToDisplay`, `searchString`, `node`, `expanded`,  I always need to set the default background color for the newly displayed nodes, because they can not be selected by definition.
The only problem can be changing the `lang`, the selected class will reset its color.
Otherwise there will be no random discharges. I do not know how to get around, I did not think about other solutions. Can you advise something?